Compiler/FrontEnd/DAEUtil.mo

```diff
@@ -4950,23 +4950,121 @@ algorithm
+        algorithm initialEquations := e :: initialEquations; then ();
+      case DAE.INITIAL_IF_EQUATION()
+        algorithm initialEquations := e :: initialEquations; then ();
+      case DAE.EQUATION()
+        algorithm equations := e :: equations; then ();
+      case DAE.EQUEQUATION()
+        algorithm equations := e :: equations; then ();
+      case DAE.ARRAY_EQUATION()
+        algorithm equations := e :: equations; then ();
+      case DAE.COMPLEX_EQUATION()
+        algorithm equations := e :: equations; then ();
+      case DAE.DEFINE()
+        algorithm equations := e :: equations; then ();
+      case DAE.ASSERT()
+        algorithm equations := e :: equations; then ();
+      case DAE.IF_EQUATION()
+        algorithm equations := e :: equations; then ();
+      case DAE.WHEN_EQUATION()
+        algorithm equations := e :: equations; then ();
+      case DAE.REINIT()
+        algorithm equations := e :: equations; then ();
+      case DAE.NORETCALL()
+        algorithm equations := e :: equations; then ();
+      case DAE.INITIAL_NORETCALL()
+        algorithm initialEquations := e :: initialEquations; then ();
+      case DAE.INITIALALGORITHM()
+        algorithm initialAlgorithms := e :: initialAlgorithms; then ();
+      case DAE.ALGORITHM()
+        algorithm algorithms := e :: algorithms; then ();
+      case DAE.CONSTRAINT()
+        algorithm constraints := e :: constraints; then ();
+      case DAE.CLASS_ATTRIBUTES()
+        algorithm classAttributes := e :: classAttributes; then ();
+      case DAE.EXTOBJECTCLASS()
+        algorithm externalObjects := e :: externalObjects; then ();
+      case DAE.COMP()
+        algorithm variables := listAppend(e.dAElist, variables); then ();
+      case DAE.FLAT_SM()
+        algorithm
+          split_comp := splitComponent(DAE.COMP(e.ident, e.dAElist,
+            DAE.emptyElementSource, SOME(SCode.COMMENT(NONE(), SOME("stateMachine")))));
+          stateMachineComps := split_comp :: stateMachineComps;
+        then
+          ();
+      case DAE.SM_COMP()
+        algorithm
+          split_comp := splitComponent(DAE.COMP(ComponentReference.crefStr(e.componentRef),
+            e.dAElist, DAE.emptyElementSource, SOME(SCode.COMMENT(NONE(), SOME("state")))));
+          stateMachineComps := split_comp :: stateMachineComps;
+        then
+          ();
+      else
+        algorithm
+          Error.addInternalError("DAEUtil.splitElements got unknown element.", Absyn.dummyInfo);
+        then
+          fail();
+    end match;
+  end for;
+
+  variables := listReverse(variables);
+  initialEquations := listReverse(initialEquations);
+  initialAlgorithms := listReverse(initialAlgorithms);
+  equations := listReverse(equations);
+  algorithms := listReverse(algorithms);
+  classAttributes := listReverse(classAttributes);
+  constraints := listReverse(constraints);
+  externalObjects := listReverse(externalObjects);
+  stateMachineComps := listReverse(stateMachineComps);
```

Comment:
you built the lists from scratch by prepending elements, so you can safely use MetaModelica.Dangerous.listReverseInPlace() here (and in the lines above)